libethereum/State.cpp

```diff
@@ -499,6 +536,50 @@ bool State::isTrieGood(bool _enforceRefs, bool _requireNoLeftOvers) const
 	return true;
 }
 
+size_t State::savepoint()
+{
+	return m_changeLog.size();
+}
+
+void State::revert(size_t _savepoint)
+{
+	while (_savepoint != m_changeLog.size())
+	{
+		auto& change = m_changeLog.back();
+		switch (change.kind)
+		{
+		case ChangeKind::storage:
+			// Public API cannot be used here because it will add another
+			// change log entry.
+//			std::cerr << "Revert storage " << change.address << "\n";
+			m_cache[change.address].setStorage(change.key, change.value);
+			break;
+		case ChangeKind::balance:
+//			std::cerr << "Revert balance " << change.address << "\n";
+			m_cache[change.address].addBalance(-change.balance);
+			break;
+		case ChangeKind::nonce:
+//			std::cerr << "Revert nonce " << change.address << "\n";
+			m_cache[change.address].setNonce(m_cache[change.address].nonce() - 1);
+			break;
+		case ChangeKind::touched:
+//			std::cerr << "Revert touch " << change.address << "\n";
+			untouch(change.address);
+			break;
+		case ChangeKind::create:
+//			std::cerr << "Revert Create " << change.address << "\n";
+			kill(change.address);
+			break;
+		case ChangeKind::prefound_create:
+//			std::cerr << "Revert prefound " << change.address << "\n";
+			setCode(change.address, {});
+			untouch(change.address);  // FIXME: add prefound and CREATE revert in single transaction.
```

Comment:
What if account was pre-funded and then created in the same transaction? Is it fine to untouch it here?